safekeeper/src/timeline.rs

```diff
@@ -598,15 +604,102 @@ impl Timeline {
+    /// do some retries anyway.
+    async fn remote_delete(&self) -> Result<()> {
+        // We will start a background task to do the deletion, so that it proceeds even if our
+        // API request is dropped.  Future requests will see the existing deletion task and wait
+        // for it to complete.
+        let mut result_rx = {
+            let mut remote_deletion_state = self.remote_deletion.lock().unwrap();
+            let result_rx = if let Some(result_rx) = remote_deletion_state.as_ref() {
+                if let Some(result) = result_rx.borrow().as_ref() {
+                    if let Err(e) = result {
+                        // A previous remote deletion failed: we will start a new one
+                        tracing::error!("remote deletion failed, will retry ({e})");
+                        None
+                    } else {
+                        // A previous remote deletion call already succeeded
+                        return Ok(());
+                    }
+                } else {
+                    // Remote deletion is still in flight
+                    Some(result_rx.clone())
+                }
+            } else {
+                // Remote deletion was not attempted yet, start it now.
+                None
+            };
+
+            match result_rx {
+                Some(result_rx) => result_rx,
+                None => self.start_remote_delete(&mut remote_deletion_state),
+            }
+        };
+
+        loop {
+            // Ignore errors, it's fine for the sender to drop, we will still see its result
+            let sender_dropped = result_rx.changed().await.is_err();
+
+            if let Some(result) = result_rx.borrow().as_ref() {
+                match result {
+                    Ok(()) => {
+                        return Ok(());
+                    }
+                    Err(e) => {
+                        return Err(anyhow::anyhow!("remote deletion failed: {e}"));
+                    }
+                }
+            } else if sender_dropped {
+                // Unexpected: we update exactly one time: this should only happen if the deletion task
+                // future is dropped without sending a result
+                return Err(anyhow::anyhow!(
+                    "remote deletion task future was dropped without sending a result"
+                ));
+            }
+        }
+    }
+
+    /// Spawn background task to do remote deletion, return a receiver for its outcome
+    fn start_remote_delete(
+        &self,
+        guard: &mut std::sync::MutexGuard<Option<RemoteDeletionReceiver>>,
+    ) -> RemoteDeletionReceiver {
+        tracing::info!("starting remote deletion");
+        let (result_tx, result_rx) = tokio::sync::watch::channel(None);
+        let ttid = self.ttid;
+        tokio::task::spawn(
+            async move {
+                let r = wal_backup::delete_timeline(&ttid).await;
```

Comment:
Does this have some kind of internal per-delete timeout? We don't want this to get stuck forever, since there is no longer any way to cancel it. Perhaps mention that in a comment.

Reply:
Thanks for checking!